src/main/kotlin/org/rust/lang/core/types/ty/TyTypeParameter.kt

```diff
@@ -16,28 +17,38 @@ data class TyTypeParameter private constructor(
     constructor(parameter: RsTypeParameter) : this(Named(parameter))
 
     constructor(trait: RsTraitItem) : this(Self(trait))
+    constructor(trait: RsTraitItem, target: String) : this(AssociatedType(trait, target))
 
     fun getTraitBoundsTransitively(): Collection<BoundElement<RsTraitItem>> =
         parameter.bounds.flatMapTo(mutableSetOf()) { it.flattenHierarchy.asSequence() }
 
+    fun getTraitRefs(): Sequence<RsTraitRef> = parameter.traitRefs
+
     override fun canUnifyWith(other: Ty, project: Project, mapping: TypeMapping?): Boolean {
         mapping?.merge(mutableMapOf(this to other))
         return true
     }
 
-    override fun substitute(map: TypeArguments): Ty = map[this] ?: this
+    override fun substitute(map: TypeArguments): Ty {
+        return map.filterKeys { it.toString() == this.toString() }.values.firstOrNull() ?: return this
```

Comment:
This reads really weird... Why `map[this]` does not work here? If it doesn't that means that `map` is constructed incorrectly, not that we should resort to approximate by name only comparison.